### ISSUE TITLE
FluentFTP.GnuTLS: Crash when create client / connect / dispose is done repeatedly

### ISSUE THREAD

**Issue by FanDjango** [robinrodricks/FluentFTP#1272]
**FTP Server OS:** Windows

**FTP Server Type:** FileZilla server

**Client Computer OS:** Windows

**FluentFTP Version:** 46.0.2

**Framework:** .NET 6

Concerns FluentFTP.GnuTLS -

Refer to #1270 - original TO @rob-hendrickx-c4t

I'm actually getting a
> System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

exception when testing with a simple (sequential) loop that does nothing more than 
* create new client
* connect
* list directory
* dispose client

So it seems it's not even an issue with thread safety? Could this be a different bug after all?

**Comment by FanDjango** [comment]
Ok, I have found the problem.

The library, which keeps all needed function pointers (delegates) in a static class, needs an explicit re-execution of the constructor code when it is used again (as in your loop).

Release 1.0.13 is out there, perhaps you would like to try with that?

**Comment by FanDjango** [comment]
And of course, also test your parallel implementation again.

**Comment by rob-hendrickx-c4t** [comment]
FYI, my parallellisation test just looks something like this:

```
[Test]
public async Task Then_we_should_be_able_to_connect_and_list_files_in_parallel()
      {
          var factory = services.GetRequiredService<ICreateCasFtpClients>();

          await Task.WhenAll(Enumerable.Repeat(1, NumberOfAttempts).Select(async _ =>
          {
              using (var client = factory.Create(new FtpCredentials
              {
                  Host = "xxx",
                  User = "xxx",
                  Password = "xxx",
                  Protocol = FtpProtocol.Ftps,
                  ValidateCertificate = false,
                  UseGnuTlsForFtps = true,
                  GnuTlsDisableAlpn = true
              }))
              {

                await client.ConnectAsync();

                var result = await client.ListDirectoryAsync("/download");
                result.Should().NotBeNull();
              }
          }));
      }
```

**Comment by rob-hendrickx-c4t** [comment]
Note that, as you can see in above snippet, every client is only used by 1 thread. But as you mentioned in the original issue, the fact that this stuff is shared statically probably is the root cause of it. So until that part is fixed, I assume the implementation is not safe for concurrent code. 

Do you happen to have a rough estimate on how long it could take before that is fixed (I fully understand it's not something that will be fixed straight away necessarily, and want to repeat that I'm amazed how fast the earlier fixes were handled)

**Comment by FanDjango** [comment]
Ok, I have looked into this.

Firstly, the entire inner structure of FluentFTP.GnuTLS is actually synchronous, as we state in the Wiki. I am overjoyed to see that it actually works well when internally "awaited" on via a `AsyncFtpClient(...)`.

So using FluentFTP.GnuTLS in an inherently async environment is not detrimental in itself. Users just need to beware of possible thread lockouts on operations that have timeouts.

GnuTLS **does** support async internally, it is just **so** much more involved (many more interop callbacks to marshal), it didn't seem worth a huge effort. Maybe someone will sit down and start coding, given the stuff that is already there as a starting point?

Why do I mention this? Well, it was the first worry when I saw your async multi thread setup.

I duplicated it and have identified a number of places where it currently fails. 

For example: I needed to serialize access to the log queue mechanism. There is only one global log queue.

But one of the other "places" is a **big** issue:

The base library wants an initial global init to setup its internal structures. After that it can be used in a thread safe fashion, for as long as you like. It then would like a global de-init to be able to dispose of its inner structures in a clean up.

Implementing this for a multi threaded use in an FTP client would mean exposing a global init (and deinit) for the user to call **before** he begins the actual use of the library as in creating ftp clients , using them and disposing them - de-init when he knows he is finished with the library.

These calls have no equivalent in an SslStream environment. FluentFTP (as the caller of FluentFTP.GnuTLS) has no knowledge of when to deinit the library. 

Currently this deinit is done when the control connection is terminated.

This whole Init/Deinit is sort of unclean and a hassle.

I have experimented just now with calling the global init on first use and completely neglecting to de-init - thereby I have successfully run 50 simultaneous threads doing a connect-disconnect with no errors.

That's a really good sign - it tells me there might be only a little bit left to tweak and bingo, problem solved.

Because when adding a GetListing call, it went "Boom" again, so this init/de-init hack would not yet be a viable (albeit "unclean") solution.

I ran into snags in the session resume logic and disposing of the credential data areas, so the whole thing would need some more work and debugging.

I really can't give you an estimate.

Linux, ASYNC (Phew, it works) **and** multi-threaded operations was sort of beyond the scope of the initial attempt to go beyond just SslStream as an option. And now look how far we have come.

I will spend some time with this to give you a better estimate, but currently I feel that there won't be a solution this week.

**Comment by FanDjango** [comment]
Ok, made a small breakthrough on this.

**If you accept** calling a _finalizer_ after the "everything" is finished, I think I have made the necessary changes to get thread-safety. 

I will work on this tomorrow and I would like to produce a branch for you (instead of a Nuget), that you could test in your environment,

I have got the Create_Client_Connect-GetListing-Disconnect_Dispose_Client litany running here with 50 enums_tasks in parallel and no hitch. Wonderful intertwined logging... Yippeee!

The only question remains: Probably will introduce a config option to decide wether the lib will dispose  and de-init on termination of the control connection or to defer until the _finalizer_ is invoked by the user. This also means, code is no longer **totally** the same between SslStream and GnuTlsStream - by the difference of one line of additional code to call the _finalizer_.

Please standby, out of the office tomorrow morning.

**Comment by FanDjango** [comment]
Closing this re. the original issue is resolved.

Follow up re parallel execution in new issue #1273
